#227 Better code completion proposal for name/key of additional properties

tfesenko opened the pull request:
This PR provides better completion proposals and code templates for properties described as additionalProperties in OAS3 and Swagger v3

# OAS3
## Code assist
`(<element> name)` is proposed for elements described with `additionalProperties` in JSON Schema. When chosen, the new name is selected for editing.
In this screencast, code assist suggests `(schema name)` for a schema and then `(property name)` for a schema describing a property:
![additionalpropertiesassist2](https://user-images.githubusercontent.com/644582/31190051-f5253f16-a907-11e7-99ae-a5f02005a22e.gif)
Similar approach applies to other components (`callbacks`, `links`, `examples`, etc) and other `additionalProperties`:
<img width="345" alt="screen shot 2017-10-04 at 1 23 35 pm" src="https://user-images.githubusercontent.com/644582/31190040-e8f915b4-a907-11e7-9b85-64063bd95927.png">
I am not attaching screenshots for all supported cases as I have provided tests for them.

### Tests
[`OpenApi3ContentAssistProcessorTest`](https://github.com/RepreZen/KaiZen-OpenAPI-Editor/pull/402/files#diff-850eaf3ece660d4c974d721e2c4a162b) provides tests for all major cases:
* Schema
* Schema Property 
* All other components (Links, Callbacks, Parameters, Examples, RequestBodies, Headers, SecuritySchemes)
* Media types
* Encodings
* Server variables

Note that some elements still have strange names:
* Discriminator#mappings: `_type_`
* OauthFlow#Scopes:  `_type_`
* Parameter names in Links: `(any name)`

Current status is covered by tests.
 
### Implementation
`JsonProposalProvider` creates proposal for the `additionalProperties` elements. The label is described by a new `TypeDefinition#getLabel()` which tried to get needed information from JSON Schema. For the schema property, we can't infer the needed name - `property` - from the schema, so I added a `title` property to JSON Schema.
### JSON Schema
I had to add one line to the JSON Schema to provide a correct `(property name)` instead of `(schema name)`
To do it, I updated our JSON Schema generator. It's generated using commit 83638080c70784ca84fcfbf86af948c6e98bcbba on
https://github.com/RepreZen/gnostic/tree/task/KZOE-227

## Code templates
Named schemas and properties now have code templates for major cases. These code templates are generated by `Activator.addNamedSchemaTemplatesInSchemas()` and `addNamedSchemaTemplatesInSchemaProperties()` using already existing code templates defined for inline schemas in `templates.xml`.
I also added code template for a schema ref for schema properties.
![additionalpropertiestemplates](https://user-images.githubusercontent.com/644582/31191057-ed1c145e-a90a-11e7-8102-41bd72b9c3f8.gif)
### Tests
* I added a new code-template context for schema properties to `SchemaObject.yaml`. It means that our JUnit tests verify that:
1. This context exists at a given location
2. All code templates provided for this context at this location will produce a valid OAS3 documentation.
### Known issues
Restore Defaults on KaiZen>OpenAPI v3/Templates preferences page erases synthetic code templates generated for named schema and named property.

### Implementation
* I added a new context for named schema properties
* `Activator.addNamedSchemaTemplatesInSchemas()` and `addNamedSchemaTemplatesInSchemaProperties()` build code templates dynamically
* Usually we use template variable to define template parts that need to be selected. Variable names don't allow spaces or special characters, and we need it to be something like `(schema name)`. So I added `ElementNameResolver` to handle it.
---------

# Swagger v2
## Code assist
Similar code assist experience was implemented for Swagger v2:
<img width="357" alt="screen shot 2017-10-04 at 2 03 24 pm" src="https://user-images.githubusercontent.com/644582/31191663-de6c1b6e-a90c-11e7-81f8-3a7a63994ed0.png">
<img width="324" alt="screen shot 2017-10-04 at 2 03 39 pm" src="https://user-images.githubusercontent.com/644582/31191664-defa13ba-a90c-11e7-96a0-a0985794ada9.png">
I provided several tests for `(schema name)` and `(property name)` in `SwaggerProposalProviderTest`, but I did not test it as thorough as foer OAS 3.

## Code templates
I did NOT provide synthetic code templates for named schemas in Swagger v2. Anyway, only one code template is available for a named schema,

## JSON Schema
Again, JSON Schema was modified to provide a correct value for `(property name)`, but as JSON Schema for Swagger v2 is not generated, there is no need to update the schema generator.

tfesenko committed:
#227 - use schema name for code-assist for additionalProperties

tfesenko committed:
#227 - code assist: select the schema name for the additionalProperties keywords

tfesenko committed:
Fix compilation errors in tests

tfesenko committed:
#227 - generate named schema templates from inline schema templates

tfesenko committed:
Merge branch 'master' into task/227

tfesenko committed:
#227 - clean up code for the 'additionalProperties' code assist, customizations for the 'property'

tfesenko committed:
Manually merge master into task/227

tfesenko committed:
Manually merge master into task/227

tfesenko committed:
#227 - use '(schema name)' as a proposal for the elements described by the 'additionalProperties'

tfesenko committed:
#227 - fix reference type provider for cases when a reference node also has a 'title'

tfesenko committed:
#227 - fix test failures in CodeTemplateTextTest (named schema templates) caused by invalid tab width pref

tfesenko committed:
#227 - add dynamic code templates for named schema properties + tests

tfesenko committed:
#227 - code templates for named schemas should create schema names in format '(schema name)' and select it for editing

tfesenko committed:
#227 - restore cursor position for empty selection - it should be at the _end_ of the inserted proposal

tfesenko committed:
#227 - tests for code assist for component keys

tfesenko committed:
#227 - more tests for code assist for the additionalProperties elements

tfesenko committed:
#227 - make getTabWidth() reusable

tfesenko committed:
#227 - remove unneeded JSON Schema customizations

tfesenko committed:
#227 - Fix code assist proposal for named schema properties in Swagger v2 + tests

tfesenko committed:
#227 - regenerate JSON Schema for OAS3

Generated using commit 83638080c70784ca84fcfbf86af948c6e98bcbba on
https://github.com/RepreZen/gnostic/tree/task/KZOE-227

tfesenko committed:
Merge branch 'master' into task/227

tfesenko committed:
Merge branch 'master' into task/227

tfesenko committed:
#227 - fix tests for code-assist for media types - they don't need a (mediaType name) value

tfesenko committed:
Merge branch 'master' into task/227

ghillairet commented on com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/schema/ComplexTypeDefinition.java:
```diff
@@ -139,4 +140,9 @@ public boolean validate(AbstractNode valueNode) {
+    
+    @Override
+    public String getLabel() {
+        return Iterables.getFirst(complexTypes, null).getLabel();
```
You will get a NPE if complexTypes is null here.

tfesenko replied:
Thanks! Addressed in  9f2cd7d

ghillairet reviewed (approved):
Passed QA, all tests passed as well.

I just left a comment about a possible NPE. 

Also I noted that there is a content assist proposal `_key_` present in  responses > content > schema. Not sure why it is there.

![screen shot 2017-10-06 at 15 19 39](https://user-images.githubusercontent.com/148045/31279734-63f72596-aaaa-11e7-8354-259e0988e91d.png)

tfesenko committed:
#227 - test that an anonymous schema in media types does not have a '_key_' proposal in code assist

tfesenko committed:
#227 - fix a potential NPE

ghillairet committed:
[#227] Better code completion proposal for name/key of additional properties

Add test to show that _key_ is still present in content assist for response content schema.

tfesenko committed:
Merge branch 'master' into task/227
